presto-main/src/main/java/com/facebook/presto/operator/aggregation/GenericAccumulatorFactory.java

```diff
@@ -100,4 +142,165 @@ public GroupedAccumulator createGroupedIntermediateAccumulator()
+
+        @Override
+        public void evaluateIntermediate(BlockBuilder blockBuilder)
+        {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public void evaluateFinal(BlockBuilder blockBuilder)
+        {
+            pagesIndex.sort(orderByChannels, orderings);
+            Iterator<Page> pagesIterator = pagesIndex.getPages(orderByOutputChannels);
+            pagesIterator.forEachRemaining(accumulator::addInput);
+            accumulator.evaluateFinal(blockBuilder);
+        }
+    }
+
+    private static class OrderingGroupedAccumulator
+            implements GroupedAccumulator
+    {
+        private final GroupedAccumulator accumulator;
+        private final List<Integer> orderByChannels;
+        private final List<SortOrder> orderings;
+        private final int[] orderByOutputChannels;
+        private final PagesIndex pagesIndex;
+
+        private OrderingGroupedAccumulator(
+                GroupedAccumulator accumulator,
+                List<Type> aggregationSourceTypes,
+                List<Integer> orderByChannels,
+                List<SortOrder> orderings,
+                PagesIndex.Factory pagesIndexFactory)
+        {
+            this.accumulator = requireNonNull(accumulator, "accumulator is null");
+            this.orderByOutputChannels = new int[aggregationSourceTypes.size()];
+            Arrays.setAll(orderByOutputChannels, IntUnaryOperator.identity());
+            this.orderByChannels = ImmutableList.copyOf(requireNonNull(orderByChannels, "orderByChannels is null"));
+            this.orderings = ImmutableList.copyOf(requireNonNull(orderings, "orderings is null"));
+            this.pagesIndex = pagesIndexFactory.newPagesIndex(aggregationSourceTypes, 10_000);
+        }
+
+        @Override
+        public long getEstimatedSize()
+        {
+            return pagesIndex.getEstimatedSize().toBytes() + accumulator.getEstimatedSize();
+        }
+
+        @Override
+        public Type getFinalType()
+        {
+            return accumulator.getFinalType();
+        }
+
+        @Override
+        public Type getIntermediateType()
+        {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public void addInput(GroupByIdBlock groupIdsBlock, Page page)
+        {
+            pagesIndex.addPage(page);
+        }
+
+        @Override
+        public void addIntermediate(GroupByIdBlock groupIdsBlock, Block block)
+        {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public void evaluateIntermediate(int groupId, BlockBuilder output)
+        {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public void evaluateFinal(int groupId, BlockBuilder output)
+        {
+            accumulator.evaluateFinal(groupId, output);
+        }
+
+        @Override
+        public void processOrderBy(GroupByHash groupByHash)
```

Comment:
I don't think this method belongs here. It makes the contract for Accumulator really weird (e.g., what's the difference/relationship between addInput and processOrderBy?). Also, it should be unaware of what a GroupByHash is -- that's an abstraction handled by the hash aggregation operator. It should probably go there, instead (i.e., the aggregation operator).

Reply:
I agree that it's strange. But I can't think of a better solution. Let's talk in person.